redpanda/resources/cluster/resource_cluster.go

```diff
@@ -558,3 +677,92 @@ func gcpConnectConsumerStructToModel(accept []*controlplanev1beta2.GCPPrivateSer
 	}
 	return output
 }
+
+func toMtlsModel(ctx context.Context, mtls *controlplanev1beta2.MTLSSpec) (*models.Mtls, diag.Diagnostics) {
+	if isMtlsSpecNil(mtls) {
+		return nil, nil
+	}
+
+	capem, err := types.ListValueFrom(ctx, types.StringType, mtls.GetCaCertificatesPem())
+	if err != nil {
+		return nil, err
+	}
+	maprules, err := types.ListValueFrom(ctx, types.StringType, mtls.GetPrincipalMappingRules())
+	if err != nil {
+		return nil, err
+	}
+	return &models.Mtls{
+		Enabled:               types.BoolValue(mtls.GetEnabled()),
+		CaCertificatesPem:     capem,
+		PrincipalMappingRules: maprules,
+	}, nil
+}
+
+func toMtlsSpec(mtls *models.Mtls) *controlplanev1beta2.MTLSSpec {
+	if isMtlsStructNil(mtls) {
+		return emptyMtlsSpec()
+	}
+	return &controlplanev1beta2.MTLSSpec{
+		Enabled:               mtls.Enabled.ValueBool(),
+		CaCertificatesPem:     utils.TypeListToStringSlice(mtls.CaCertificatesPem),
+		PrincipalMappingRules: utils.TypeListToStringSlice(mtls.PrincipalMappingRules),
+	}
+}
+
+func isMtlsNil(container any) bool {
+	v := reflect.ValueOf(container)
+	if !v.IsValid() || v.IsNil() {
+		return true
+	}
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+	if v.Kind() != reflect.Struct {
+		return true
+	}
+	mtlsField := v.FieldByName("Mtls")
+	if !mtlsField.IsValid() || mtlsField.IsNil() {
+		return true
+	}
+	return isMtlsStructNil(mtlsField.Interface().(*models.Mtls))
+}
+
+func isMtlsStructNil(m *models.Mtls) bool {
+	return m == nil || (m.Enabled.IsNull() && m.CaCertificatesPem.IsNull() && m.PrincipalMappingRules.IsNull())
+}
+
+func isMtlsSpecNil(m *controlplanev1beta2.MTLSSpec) bool {
+	return m == nil || (!m.GetEnabled() && len(m.GetCaCertificatesPem()) == 0 && len(m.GetPrincipalMappingRules()) == 0)
+}
+
+func emptyMtlsSpec() *controlplanev1beta2.MTLSSpec {
+	return &controlplanev1beta2.MTLSSpec{
+		Enabled:               false,
+		CaCertificatesPem:     make([]string, 0),
+		PrincipalMappingRules: make([]string, 0),
+	}
+}
+
+func isAwsPrivateLinkStructNil(m *models.AwsPrivateLink) bool {
+	return m == nil || (m.Enabled.IsNull() && m.ConnectConsole.IsNull() && m.AllowedPrincipals.IsNull())
```

Comment:
[question, to all the handlers below]: if m is not nil, and the model's content is required (L85 resourceClusterSchema and below), how can we have non-nil maps with required content empty?

Reply:
I ran into it as an issue with testing so I wanted to be very thorough in testing for nil or empty. There's a decent chance it's excessive. 

As far as potential scenarios, malformed manual edits on state file would be one scenario where fields could be populated but with nil values which probably should be discarded. Hand edits of the state file are pretty rare in a post tf 0.12 world but I've had to do it enough times that I'd like the additional certainty just in case.